### ISSUE TITLE
Eigenvectors are missing

### ISSUE THREAD

**Issue by rtoy** [rtoy/maxima#1496]
Imported from SourceForge on 2024-07-04 17:07:44
Created by **[pbruin](https://sourceforge.net/u/pbruin/)** on 2015-08-28 22:04:57
Original: https://sourceforge.net/p/maxima/bugs/3008

---

It seems that the fix for bug #2183 introduced a new bug causing eigenvectors() to miss some eigenvectors:

    Maxima branch_5_37_base_17_gce242fd http://maxima.sourceforge.net
    using Lisp GNU Common Lisp (GCL) GCL 2.6.12
    Distributed under the GNU Public License. See the file COPYING.
    Dedicated to the memory of William Schelter.
    The function bug_report() provides bug reporting information.
    (%i1) M: matrix([1, 1, 0], [0, 1, 0], [0, 0, 2]);
                                      [ 1  1  0 ]
                                      [         ]
    (%o1)                             [ 0  1  0 ]
                                      [         ]
                                      [ 0  0  2 ]
    (%i2) eigenvectors(M);
    (%o2)                  [[[1, 2], [2, 1]], [[[1, 0, 0]]]]

The eigenvector [0, 0, 1] is missing.  After reverting commit dcd222b (I'm not suggesting this is the right fix), the correct output is given:

    (%i2) eigenvectors(M);
    (%o2)           [[[1, 2], [2, 1]], [[[1, 0, 0]], [[0, 0, 1]]]]

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:07:46
Created by **[mitreude](https://sourceforge.net/u/mitreude/)** on 2015-12-20 17:17:07
Original: https://sourceforge.net/p/maxima/bugs/3008/#bd6d

---

I agree that the patch for bug #2183 seems to be the cause of the issue.  As an experiment, I suggest eliminating (or commenting out) the line

~~~~
count:length(multiplicities),
~~~~

and replacing it with the line

~~~~
count:apply("+",multiplicities),
~~~~

in the patched file "eigen.mac".  In my limited testing, that change seemed to fix things for me.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:07:49
Created by **[pbruin](https://sourceforge.net/u/pbruin/)** on 2015-12-22 15:09:16
Original: https://sourceforge.net/p/maxima/bugs/3008/#bd6d/d196

---

This certainly looks like the correct fix to me.  (I also checked that it fixes my example, but haven't done extensive tests.)

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:07:53
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2016-05-17 00:22:44
Original: https://sourceforge.net/p/maxima/bugs/3008/#bd6d/00c9

---

Re: `count:apply("+",multiplicities)`, I'm pretty sure that's incorrect. In the loop at line 121 in share/matrix/eigen.mac, `index1` is supposed to be an index for the eigenvalues and multiplicities, which are the same length. I'm pretty sure that it can't be right to assign anything other than `length(multiplicities)` (as it is at present).

I have a different patch which I am working on. I'll try to update this report soon.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:07:57
Created by **[mitreude](https://sourceforge.net/u/mitreude/)** on 2016-05-18 15:12:48
Original: https://sourceforge.net/p/maxima/bugs/3008/#bd6d/00c9/7057

---

My suggestion of `count:apply("+",multiplicities)` was based on a cursory examination of the code in the #2183 patch only.  From the standpoint of not changing code that was in place prior to #2183, I believe my suggestion was correct.  With that being said, my current thinking is that a better solution is to keep `count:length(multiplicities)` and eliminate `count:mi(count-part(multiplicities,index1)+1)` in line 122.

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:08:00
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2016-05-22 07:00:42
Original: https://sourceforge.net/p/maxima/bugs/3008/#e47e

---

- **status**: open --> closed

**Comment by rtoy** [comment]
Imported from SourceForge on 2024-07-04 17:08:04
Created by **[robert_dodier](https://sourceforge.net/u/robert_dodier/)** on 2016-05-22 07:00:42
Original: https://sourceforge.net/p/maxima/bugs/3008/#f313

---

Fixed by commit 3e4e107. Closing this report.
